### PR TITLE
Fix Adventures exiling themselves before applying their effects

### PR DESCRIPTION
**Pull request by Susucre**
The Adventure were adding their exile-on-resolve effect first in the mode's effect list.
This was an issue with giving abilities to the spell on the stack (like the lifelink from Heartflame Duelist). Added a finalizeAdventure() method to add it after the spell has been set up.

~~The card is not yet official, but here is the implemented text:~~

> Heartflame Duelist {1}{W}
> Creature — Human Knight
>
> Instant and sorcery spells you control have lifelink.
>
> 3/1
>
> //
> Heartflame Slash {2}{R}
> Instant — Adventure
>
> Heartflame Slash deals 3 damage to any target.

### CODE REVIEW

**Comment by Susucre** [comment]
Has been officially spoiled today.

**Comment by theelk801** [comment]
can you break the refactor part into a separate PR?

**Comment by Susucre** [comment]
I need to check, but the first commit is for the card alone, and the rest is for the fix on all adventures. Will rebase/split.

**Comment by Susucre** [comment]
#11032 has the simple implementation, bugged with itself (with an Heartflame Duelist in play, Heartflame Slash does not gain life, Lightning Bolt does.)

File: Mage/src/main/java/mage/cards/AdventureCardSpellImpl.java
Comment (JayDi85) [inline Mage/src/main/java/mage/cards/AdventureCardSpellImpl.java:110]: finalized = false

File: Mage/src/main/java/mage/cards/AdventureCardSpellImpl.java
Comment (JayDi85) [inline Mage/src/main/java/mage/cards/AdventureCardSpellImpl.java:122]: Try to mark developer related errors like "Wrong code usage. xxx."

Example: Wrong code usage. Adventures must call finalizeAdventure() at the end of card's setup.

**Review by JayDi85 — APPROVED** [review]

**Comment by Susucre** [comment]
@JayDi85 the Heartflame Duelist will be added in #11032 at theelk801's request and this branch is based on top of that. I think the Duelist branch may be merged soon, and this one can be delayed for a little while. Hence the name change.

**Comment by xenohedron** [comment]
Let's wait until all other adventures are merged, then rebase against master and make sure all of them are fixed, then merge this.

**Comment by xenohedron** [comment]
Looks like all adventures have been implemented

**Comment by Susucre** [comment]
Finalized them all. All clear to merge.

**Pull request merged**
